dspy/predict/program_of_thought.py

```diff
@@ -0,0 +1,70 @@
+import dsp
+import dspy
+from ..primitives.program import Module
+from ..primitives.interpreter import CodePrompt, PythonInterpreter
+import re
+
+class ProgramOfThought(Module):
+    def __init__(self, signature, max_iters=3):
+        super().__init__()
+
+        self.signature = signature = dspy.Predict(signature).signature
+        self.max_iters = max_iters
+
+        self.input_fields = {k: v for k, v in signature.kwargs.items() if isinstance(v, dspy.InputField)}
+        self.output_fields = {k: v for k, v in signature.kwargs.items() if isinstance(v, dspy.OutputField)}
+
+        inputs, outputs = signature.fields[:-1], signature.fields[-1:]
+
+        inputs_ = ', '.join([f"`{field.input_variable}`" for field in inputs])
```

Comment:
Okay ideally this parsing doesn't need to happen. Ideally this uses your new signature things.

Reply:
Basically all signature manipulations here should be "clean". Feel free to close this PR and basically make it part of the other one if that helps